crates/sodium/src/crypto_system/secure_buffer.rs

```diff
@@ -0,0 +1,131 @@
+use lib3h_crypto_api::{Buffer, BufferType, CryptoError, CryptoResult, ProtectState};
+
+use crate::check_init;
+use libc::c_void;
+
+/// A secure buffer implementation of lib3h_crypto_api::Buffer
+/// making use of libsodium's implementation of mlock and mprotect.
+pub struct SecureBuffer {
+    z: *mut c_void,
+    s: usize,
+    p: std::cell::RefCell<ProtectState>,
+}
+
+unsafe impl Send for SecureBuffer {}
+
+impl Drop for SecureBuffer {
+    fn drop(&mut self) {
+        unsafe {
+            rust_sodium_sys::sodium_free(self.z);
+        }
+    }
+}
+
+impl Clone for SecureBuffer {
+    fn clone(&self) -> Self {
+        let mut out = SecureBuffer::new(self.s).expect("could not alloc new");
```

Comment:
This looks like a runtime problem.  Shouldn't we not be doing an `expect` here?

Reply:
Alas, clone must return Self, not a result, so we have to panic!... But rust panics on all memory allocation errors, so it's consistent at least.